hw/xbox/nv2a/nv2a.c

```diff
@@ -507,10 +507,21 @@ static void nv2a_realize(PCIDevice *dev, Error **errp)
     }
 
     /* init fifo cache1 */
+    qemu_spin_init(&d->pfifo.cache1.alloc_lock);
     qemu_mutex_init(&d->pfifo.cache1.cache_lock);
     qemu_cond_init(&d->pfifo.cache1.cache_cond);
     QSIMPLEQ_INIT(&d->pfifo.cache1.cache);
     QSIMPLEQ_INIT(&d->pfifo.cache1.working_cache);
+    QSIMPLEQ_INIT(&d->pfifo.cache1.available_entries);
+    QSIMPLEQ_INIT(&d->pfifo.cache1.retired_entries);
+
+    /* Pre-allocate memory for CacheEntry objects */
+    for (i=0; i < 100000; i++) {
```

Comment:
Possibly overkill but the structure is small, and a frame in MP of Halo:CE uses ~50k.

Reply:
For what it's worse, the real Xbox has CACHE1 which holds 128 elements (method + data pair).
This actually important, because that means a method causing an interrupt being hit guarantees that a command in 128+ methods, has not been loaded yet, and can still be modified.
If I remember correctly, we had cases where the games were unable to inject further commands, because XQEMU pretended to have loaded *all* of them already.

I think our bottleneck would be 0x1800 methods which are used during draw calls (easily a couple of thousand per draw), so our cache has to be fast - and I'm not sure how fast it is right now.

If we do this properly, we'd even use the CACHE1 registers (array) in PGRAPH for this, no longer a list (how does QEMU even implement it?).
However, the CACHE1 behaves.. strangely. I've been experimenting with it for [nv2a-trace](https://github.com/XboxDev/nv2a-trace), and I couldn't really figure out how it works (the data looked good, but it's PUT and GET pointers were... odd). Also issues where the entire CACHE1 has the same method+data pair 128 times suddenly.
More research necessary (or doc reading?!).